### PR TITLE
Use stale information if it's available to answer requests quickly

### PR DESCRIPTION
**Pull request by mpickering**
With this patch whenever any request made by the user is dealt with by an `IdeAction` which has the implicit guarantee to never block for an unbounded amount of time. 

Stale information is used if it's available to answer all requests. Information should usually be up-to-date because the `FileStore.hs` rule requests up-to-date information whenver a file changes. 

Do not review the changes to the hover/completions etc code too closely because that is all rewritten on another branch by Zubin anyway to use HieDb. 

This patch also lays the way for removing `kick` and more fine grained control over which files are typechecked.

### CODE REVIEW

File: bench/Main.hs [hunk 1]
```diff
@@ -218,14 +218,16 @@ runBenchmarks benchmarks = do
     putStrLn $ intercalate ", " $ map (showDuration . snd) results
```
Comment (ndmitchell) [inline bench/Main.hs:218]: How big a deal would it be to send the benchmark changes in via a different PR? Seems totally unrelated to the rest of it, and one for @pepeiborra to review.

File: bench/Main.hs [hunk 2]
```diff
@@ -218,14 +218,16 @@ runBenchmarks benchmarks = do
...
 
 runBench :: HasConfig => Bench -> IO Seconds
-runBench Bench {..} = handleAny (\e -> print e >> return (-1))
+runBench Bench {..} = handleAny (\e -> print e >> return (-2))
```
Comment (ndmitchell) [inline bench/Main.hs:221]: Having -1 as a special case was odd, but just on the edge of tolerable. Now we're having two magic negative seconds I think an enum would be better.

File: bench/Main.hs [hunk 3]
```diff
@@ -236,17 +238,15 @@ runBench Bench {..} = handleAny (\e -> print e >> return (-1))
...
-    liftIO $ threadDelay 50000
-
     return $ if res then t else -1
   where
     cmd =
       unwords $
         [ "ghcide",
           "--lsp",
+          "--test",
+          "--verbose",
```
Comment (ndmitchell) [inline bench/Main.hs:249]: Why pass verbose? Why not take this as an argument to get passed onward?

File: bench/Main.hs [hunk 4]
```diff
@@ -236,17 +238,15 @@ runBench Bench {..} = handleAny (\e -> print e >> return (-1))
...
-    -- sleeep to give ghcide a chance to print the RTS stats
-    liftIO $ threadDelay 50000
-
     return $ if res then t else -1
   where
     cmd =
       unwords $
         [ "ghcide",
           "--lsp",
+          "--test",
```
Comment (ndmitchell) [inline bench/Main.hs:248]: Why is this required?

File: src/Development/IDE/Core/Service.hs [hunk 5]
```diff
@@ -50,7 +47,7 @@ initialise :: LSP.ClientCapabilities
            -> IdeOptions
            -> VFSHandle
            -> IO IdeState
-initialise caps mainRule getLspId toDiags logger debouncer options vfs =
+initialise caps mainRule getLspId toDiags logger debouncer options vfs = do
```
Comment (ndmitchell) [inline src/Development/IDE/Core/Service.hs:50]: What's this change for?

File: src/Development/IDE/Core/Shake.hs [hunk 6]
```diff
@@ -419,45 +457,113 @@ withMVar' var unmasked masked = mask $ \restore -> do
...
+pattern DelayedAction :: String ->  Logger.Priority -> Action a -> DelayedAction a
+pattern DelayedAction a b c = DelayedActionX a b c
+
+mkDelayedAction :: String -> Logger.Priority -> Action a -> DelayedAction a
+mkDelayedAction = DelayedAction
+
+data DelayedActionX a = DelayedActionX { actionName :: String -- Name we show to the user
+                                      -- The queue only contrains entries for
+                                      -- unique key values.
+                                      , actionPriority :: Logger.Priority
```
Comment (ndmitchell) [inline src/Development/IDE/Core/Shake.hs:487]: Why are we using log priorities here?

File: src/Development/IDE/Core/Shake.hs [hunk 7]
```diff
@@ -524,6 +634,30 @@ newSession IdeState{shakeExtras=ShakeExtras{..}, ..} systemActs userActs = do
...
+logDelayedAction :: Logger -> DelayedActionInternal -> Action ()
+logDelayedAction l d  = do
+    start <- liftIO $ offsetTime
+    getAction d
+    runTime <- liftIO $ start
+    return ()
+    liftIO $ logPriority l (actionPriority d) $ T.pack $
+        "finish: " ++ (actionName d) ++ " (took " ++ showDuration runTime ++ ")"
+
+freshId :: Var Int -> IO Int
```
Comment (ndmitchell) [inline src/Development/IDE/Core/Shake.hs:657]: Why not just reuse Data.Unique?

File: src/Development/IDE/Core/Shake.hs [hunk 8]
```diff
@@ -524,6 +634,30 @@ newSession IdeState{shakeExtras=ShakeExtras{..}, ..} systemActs userActs = do
...
+freshId qcount = do
+    modifyVar qcount (\n -> return (n + 1, n))
```
Comment (ndmitchell) [inline src/Development/IDE/Core/Shake.hs:659]: Looks like a potential space leak.

File: src/Development/IDE/Plugin/Completions/Logic.hs [hunk 9]
```diff
@@ -395,7 +395,6 @@ getCompletions ideOpts CC { allModNamesAsNS, unqualCompls, qualCompls, importabl
         = filtModNameCompls ++ map (toggleSnippets caps withSnippets
                                       . mkCompl ideOpts . stripAutoGenerated) filtCompls
                             ++ filtKeywordCompls
-
```
Comment (ndmitchell) [inline src/Development/IDE/Plugin/Completions/Logic.hs:398]: Why delete one line in this file?

File: bench/Main.hs (hunk 1 above)
Comment (pepeiborra) [reply bench/Main.hs:218]: Agreed. I've got a number of changes for the benchmark suite myself, so a separate PR that can be reviewed and merged quickly would be much appreciated.

**Comment by pepeiborra** [comment]
Benchmarks please!

**Comment by mpickering** [comment]
Benchmarks are vacuous now but anyway.

```
TOTAL hover = 0.01s (10 repetitions)
TOTAL getDefinition = 0.03s (10 repetitions)
TOTAL documentSymbols = 0.65s (100 repetitions)
TOTAL documentSymbols after edit = 0.60s (100 repetitions)
TOTAL completions after edit = 0.16s (10 repetitions)
TOTAL code actions = 0.02s (10 repetitions)
TOTAL code actions after edit = 7.98s (10 repetitions)
Benchmark ghcide-bench: FINISH
```

File: bench/Main.hs (hunk 1 above)
Comment (pepeiborra) [reply bench/Main.hs:218]: I've pushed my benchmarks changes in #625

File: bench/Main.hs (hunk 4 above)
Comment (wz1000) [reply bench/Main.hs:248]: See https://github.com/digital-asset/ghcide/pull/624/files#diff-9f3c72df87c6e40a0ab15d33001d156cR739

**Comment by mpickering** [comment]
I am taking a week break so probably won't get back to this until after that.

**Comment by wz1000** [comment]
>     4. I’m slightly worried about the fact that we switch to a different codepath for testing in `useWithStaleFast'`. I understand why but it still makes it easy for things to creep in that aren’t covered by testing.

I would like to follow the codepath we follow for tests even when not testing, but @mpickering disagrees.

I think it is OK to block on the first request when no results have been computed. Otherwise you get some very weird behavior.

For example, suppose you have an `MaybeT IdeAction` which uses `n` rules, where none of the rules have been computed before. When you run the action the first time, it will fail after trying to get the results of the first rule, and queue up a request to compute the rule. The next time, it will fail when `useE`ing the second rule(if the first rule has finished being computed), and so on. In the best case scenario, the action will have to run at least `n` times before it succeeds.

File: src/Development/IDE/Core/Shake.hs [hunk 10]
```diff
@@ -407,12 +451,72 @@ withMVar' var unmasked masked = mask $ \restore -> do
     putMVar var a'
     pure c
 
+
+-- | Actions with an ID for tracing purposes
+data DelayedActionExtra
+  = DelayedActionExtra
+  { _actionInternalId :: Unique
```
[the remark this answers is not in the document]
Comment (wz1000) [reply src/Development/IDE/Core/Shake.hs:458]: I believe it is meant to be used for tracing purposes, maybe it will be helpful with the work @mpardalos is going to do.

**Comment by wz1000** [comment]
> I’m less happy to do the same for things not triggered by users even if they are triggered by their editor. This includes things like codelenses, document outline and so on. There are two reasons why I don’t like it here:

One problem with this is that vscode does not send any new requests until it gets a response for these. To the user, this looks the server blocking on hover, go to definition etc.

This means that in practice, a lot of the gains of this PR will be negated. For instance, on modification, vscode will send these requests. Then when the user tries to hover on something, vscode will refuse to send any new requests until the code lenses etc. triggered by the modification return.

I don't really have a problem with this though since I don't use vscode and my editor only sends outline requests when I ask it to.

**Comment by pepeiborra** [comment]
> > I’m less happy to do the same for things not triggered by users even if they are triggered by their editor. This includes things like codelenses, document outline and so on. There are two reasons why I don’t like it here:
> 
> One problem with this is that vscode does not send any new requests until it gets a response for these. To the user, this looks the server blocking on hover, go to definition etc.
> 
> This means that in practice, a lot of the gains of this PR will be negated. For instance, on modification, vscode will send these requests. Then when the user tries to hover on something, vscode will refuse to send any new requests until the code lenses etc. triggered by the modification return.
> 
> I don't really have a problem with this though since I don't use vscode and my editor only sends outline requests when I ask it to.

Are you positive that it is VSCode who doesn't send the requests concurrently? 
I believe that ghcide currently deals with requests single-threadedly, so I don't this is VSCode specific.

**Comment by wz1000** [comment]
> Are you positive that it is VSCode who doesn't send the requests concurrently?
> I believe that ghcide currently deals with requests single-threadedly, so I don't this is VSCode specific.

I think so, since I got frustrated by this once and checked the logs.

**Comment by pepeiborra** [comment]
> > Are you positive that it is VSCode who doesn't send the requests concurrently?
> > I believe that ghcide currently deals with requests single-threadedly, so I don't this is VSCode specific.
> 
> I think so, since I got frustrated by this once and checked the logs.

Do you mean the logs printed by ghcide? There's every chance that those would be single-threaded too

**Comment by wz1000** [comment]
I'm not sure now, they might have been the haskell-lsp logs. I will check again when I'm at my computer.

**Comment by wz1000** [comment]
I have simplified the API somewhat by removing `delay`, `shakeRunUser` and `shakeRunInternal`.

Now we have `shakeEnqueue` and `shakeRestart`, along with `delayedAction`, which is essentially `shakeEnqueue` for the `IdeAction` monad.

**Comment by wz1000** [comment]
I believe this is ready for another look. I think I've addressed most of the comments here.

**Comment by pepeiborra** [comment]
> I'm not sure now, they might have been the haskell-lsp logs. I will check again when I'm at my computer.

Did you have a chance to check whether the lap-client is able to send requests concurrently?

**Comment by pepeiborra** [comment]
Since this branch has performance impacting changes, could you run and share differential benchmarks following the guidelines below

https://github.com/digital-asset/ghcide/blame/master/README.md#L313-L321

**Comment by alanz** [comment]
> Since this branch has performance impacting changes

Given that this PR splits requests into a fast path and a slow path, we must be careful in analysing any performance graphs we produce.

**Comment by wz1000** [comment]
> Did you have a chance to check whether the lsp-client is able to send requests concurrently?

@pepeiborra you are right, my observations were due to the single threaded event loop in ghcide. Now I'm wondering whether we should make that fork. 

Certainly making code actions synchronous now will negate many of the benefit of this PR in vscode.

**Comment by wz1000** [comment]
@pepeiborra I get the following:

```
❯ stack --stack-yaml=stack88.yaml build ghcide:benchHist && stack --stack-yaml=stack88.yaml exec benchHist
# ghcide-bench (for bench-hist/upstream/documentSymbols_after_edit.csv bench-hist/upstream/documentSymbols_after_edit.benchmark-gcStats)
Error when running Shake build system:
  at want, called at bench/hist/Main.hs:92:3 in main:Main
* Depends on: all
  at need, called at bench/hist/Main.hs:113:7 in main:Main
* Depends on: bench-hist/upstream/documentSymbols_after_edit.svg
  at readFileLines, called at bench/hist/Main.hs:380:10 in main:Main
  at readFile', called at src/Development/Shake/Internal/Derived.hs:131:30 in shake-0.18.5-Lg6S0PWFuV673fv8WVPiwb:Development.Shake.Internal.Derived
  at need, called at src/Development/Shake/Internal/Derived.hs:118:15 in shake-0.18.5-Lg6S0PWFuV673fv8WVPiwb:Development.Shake.Internal.Derived
* Depends on: bench-hist/upstream/documentSymbols_after_edit.benchmark-gcStats
  at &%>, called at bench/hist/Main.hs:180:5 in main:Main
* Depends on: bench-hist/upstream/documentSymbols_after_edit.csv bench-hist/upstream/documentSymbols_after_edit.benchmark-gcStats
* Raised the exception:
ghcide-bench: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
```

**Comment by pepeiborra** [comment]
benchHist needs a `ghcide-bench` binary in the path to work. It would be nice to automate this, but I tried a few things and couldn't get them to work.

You can manually change the path in the `hist.yaml` file

**Comment by wz1000** [comment]
Is `ghcide-bench` the usual `ghcide` executable? If I change `ghcide-bench` to `ghcide` in `hist.yaml`, I get:

```
Command line: ghcide --timeout=3000 --samples=100 --csv=bench-hist/upstream/documentSymbols_after_edit.csv --example-package-version=3.0.0.0 --rts=-I0.5 --ghcide=bench-hist/upstream/ghcide --select 'documentSymbols after edit' -q
Exit code: 1
Stderr:
Invalid option `--timeout=3000'
```

**Comment by pepeiborra** [comment]
`ghcide-bench` is the benchmark executable

**Comment by alanz** [comment]
> Certainly making code actions synchronous now will negate many of the benefits of this PR in vscode.

What do you mean by this? Only about benchmark results? I see plenty of overlapping requests from vscode, that would benefit from using cached data.

**Comment by pepeiborra** [comment]
> @pepeiborra you are right, my observations were due to the single threaded event loop in ghcide. Now I'm wondering whether we should make that fork.
> 
> Certainly making code actions synchronous now will negate many of the benefits of this PR in vscode.

I am confused too, I thought this PR didn't make code action asynchronous. Does it?

It's good to have confirmation that it is ghcide that enforces a sequential lsp request handling. We should consider addressing that as an alternative to using stale data for everything.

**Comment by wz1000** [comment]
Sorry, I meant that not using stale data for code actions while maintaining a synchronous request handling would negate many of the benefits of this PR. This is because vscode frequently sends code action requests, which depend on `TypeCheck`. This means we cannot answer any hover requests while ghcide is computing the response to the code action request. Ideally, we would be able to answer the hover request immediately with stale data.

Asynchronous request processing is not a alternative for this PR since individual requests would still have to wait for data to be recomputed. I think it would, however, be OK to use asynchronous processing for things like code actions to ensure that we don't block for requests like hover.

**Comment by pepeiborra** [comment]
But this PR doesn't change how code actions work, they are still using Shake data, right? This is what confuses me.

**Comment by wz1000** [comment]
Ah, I meant code lens. Code action doesn't use `Typecheck` any way, so it shouldn't be much of a problem.

**Comment by wz1000** [comment]
Benchmark results:

```
name                       | success | samples | startup | setup | experiment | maxResidency
-------------------------- | ------- | ------- | ------- | ----- | ---------- | ------------
hover                      | True    | 10      | 6.98s   | 0.00s | 0.05s      | 57MB        
edit                       | True    | 10      | 6.73s   | 0.00s | 2.90s      | 192MB       
hover after edit           | True    | 10      | 6.77s   | 0.00s | 0.26s      | 57MB        
getDefinition              | True    | 10      | 6.95s   | 0.00s | 0.07s      | 56MB        
documentSymbols            | True    | 100     | 7.09s   | 0.00s | 0.58s      | 56MB        
documentSymbols after edit | True    | 100     | 6.74s   | 0.00s | 1.14s      | 197MB       
completions after edit     | True    | 10      | 6.67s   | 0.00s | 0.56s      | 57MB        
code actions               | True    | 10      | 6.91s   | 0.24s | 0.11s      | 57MB        
code actions after edit    | True    | 10      | 6.51s   | 0.00s | 3.61s      | 219MB       
```

**Comment by wz1000** [comment]
Here is the entire output of the benchmark tool:

```
Running hover benchmark
0.04s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
Running edit benchmark
0.24s
0.24s
0.59s
0.35s
0.22s
0.22s
0.38s
0.22s
0.22s
0.22s
Running hover after edit benchmark
0.24s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
Running getDefinition benchmark
0.05s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
Running documentSymbols benchmark
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.01s
0.01s
0.01s
0.02s
0.00s
0.01s
0.01s
0.01s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.01s
0.02s
0.01s
0.01s
0.01s
0.01s
0.00s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.00s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.00s
0.01s
0.01s
0.01s
0.00s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.00s
0.00s
0.00s
0.01s
0.01s
0.00s
0.01s
0.01s
0.00s
0.00s
0.00s
0.02s
0.01s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.00s
0.01s
0.01s
0.00s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.02s
0.00s
0.00s
0.00s
0.00s
0.00s
Running documentSymbols after edit benchmark
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.02s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.16s
0.01s
0.01s
0.01s
0.01s
0.03s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.02s
0.01s
0.19s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.01s
0.03s
0.01s
0.01s
0.01s
Running completions after edit benchmark
0.26s
0.01s
0.02s
0.02s
0.02s
0.02s
0.02s
0.02s
0.18s
0.01s
Running code actions benchmark
0.08s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
0.00s
Running code actions after edit benchmark
0.30s
0.21s
0.38s
0.19s
0.17s
0.48s
0.59s
0.38s
0.53s
```

As we can see, for the results returning stale information, there is a pause for the initial run. Subsequent requests return instantaneously.

**Comment by wz1000** [comment]
Here are the differential results: http://78.47.113.11/results/

**Comment by wz1000** [comment]
I think the outstanding questions that remain are:

1. Should we block on the initial computation of a request (like we do when `--testing`) or fail until we have a cache? 
2. Should we respond immediately with stale information for requests like `codeLens`

**Comment by pepeiborra** [comment]
1. I don't enjoy having to hover twice, go-to-def twice, etc, so I'd prefer blocking here.
2. It seems trivial to make ghcide handle requests concurrently, see https://github.com/pepeiborra/ghcide/tree/multi-threaded, so I'd prefer to be synchronous for code lenses and all other editor triggered requests

**Comment by wz1000** [comment]
> * I don't enjoy having to hover twice, go-to-def twice, etc, so I'd prefer blocking here.
> 
> * It seems trivial to make ghcide handle requests concurrently, see https://github.com/pepeiborra/ghcide/tree/multi-threaded, so I'd prefer to be synchronous for code lenses and all other editor triggered requests

Addressed both points.

**Comment by wz1000** [comment]
From the differential benchmarks, this result seems to be the most important/striking one:

**Comment by pepeiborra** [comment]
> From the differential benchmarks, this result seems to be the most important/striking one:
> 
> 

Very nice!! That's much better than the optimisation in #670, which is bottle-necked a bit by Shake

**Comment by wz1000** [comment]
See also

File: src/Development/IDE/Core/Shake.hs [hunk 11]
```diff
@@ -620,6 +696,55 @@ usesWithStale_ key files = do
...
+
+askShake :: IdeAction ShakeExtras
+askShake = ask
+
+-- | A (maybe) stale result now, and an up to date one later
+data FastResult a = FastResult { stale :: Maybe (a,PositionMapping), uptoDate :: IO (Maybe a)  }
+
+-- | Lookup value in the database and return with the stale value immediately
+-- Will queue an action to refresh the value.
+-- Never blocks.
```
Comment (cocreature) [inline src/Development/IDE/Core/Shake.hs:716]: This is no longer accurate. You block the first time (which is a good idea imho so just update the docs).

File: src/Development/IDE/Core/Shake.hs [hunk 12]
```diff
@@ -407,12 +451,72 @@ withMVar' var unmasked masked = mask $ \restore -> do
...
+
+{-# COMPLETE DelayedAction#-}
+pattern DelayedAction :: String ->  Logger.Priority -> Action a -> DelayedAction a
+pattern DelayedAction a b c = DelayedActionX a b c
+
+mkDelayedAction :: String -> Logger.Priority -> Action a -> DelayedAction a
+mkDelayedAction = DelayedAction
+
+data DelayedActionX a = DelayedActionX
+  { actionName :: String -- ^ Name we show to the user
```
[the remark this answers is not in the document]
Comment (cocreature) [reply src/Development/IDE/Core/Shake.hs:504]: Can we clarify the doc comment here please?

**Comment by wz1000** [comment]
Fixed!
